synapse/storage/databases/main/events_worker.py

```diff
@@ -1287,3 +1295,55 @@ def get_next_event_to_expire_txn(txn):
         return await self.db_pool.runInteraction(
             desc="get_next_event_to_expire", func=get_next_event_to_expire_txn
         )
+
+    async def get_event_id_from_transaction_id(
+        self, user_id: str, token_id: str, txn_id: str
+    ) -> Optional[str]:
+        """Look up if we have already persisted an event for the transaction ID,
+        returning the event ID if so.
+        """
+        return await self.db_pool.simple_select_one_onecol(
+            table="event_txn_id",
+            keyvalues={"user_id": user_id, "token_id": token_id, "txn_id": txn_id},
+            retcol="event_id",
+            allow_none=True,
+            desc="get_event_id_from_transaction_id",
+        )
+
+    async def get_already_persisted_events(
+        self, events: Iterable[EventBase]
+    ) -> Dict[str, str]:
+        """Look up if we have already persisted an event for the transaction ID,
+        returning a mapping from event ID in the given list to the event ID of
+        an existing event.
+        """
+
+        mapping = {}
+
+        for event in events:
+            token_id = getattr(event.internal_metadata, "token_id", None)
+            txn_id = getattr(event.internal_metadata, "txn_id", None)
+            if token_id and txn_id:
+                existing = await self.get_event_id_from_transaction_id(
+                    event.sender, token_id, txn_id
+                )
```

Comment:
I don't know if we want to try and batch these up a bit? The standard `simple_select_many` doesn't work due to it only supporting one column. There are psycopg2 helper functions that might help us though

Reply:
It looks like something similar could be made which takes an iterable of tuples (or dicts)?

Reply:
Yup, I'm just not sure how much I want to try and do that for something that is going in an RC

Reply:
park it for now, optimise later?